pyro/infer/elbo.py

```diff
@@ -5,13 +5,26 @@
 import warnings
 from abc import ABCMeta, abstractmethod
 
+import torch
+
 import pyro
 import pyro.poutine as poutine
 from pyro.infer.util import is_validation_enabled
 from pyro.poutine.util import prune_subsample_sites
 from pyro.util import check_site_shape
 
 
+class _ELBOModule(torch.nn.Module):
+    def __init__(self, model: torch.nn.Module, guide: torch.nn.Module, elbo: "ELBO"):
+        super().__init__()
+        self.model = model
+        self.guide = guide
+        self.elbo = elbo
+
+    def forward(self, *args, **kwargs):
+        return self.elbo.differentiable_loss(self.model, self.guide, *args, **kwargs)
```

Comment:
Regarding your concern

 > Note that Pyro's global parameter store may cause this new interface to behave unexpectedly relative to standard PyTorch when working with :class:`~pyro.nn.PyroModule` s.

have you considered using a scope here?  I haven't thought this through, but I suppose we could use a param store scope, like this
```python
class ELBOModule(...):
    def __init__(self, ...):
        ...
        self.param_scope = {}

    def forward(self, ...):
        with pyro.params.param_store.scope(self.param_scope):
            return self.elbo.differentiable_loss(...)
```
For example, could a scope here simplify or replace the global `_PYRO_MODULE_LOCAL_PARAM` machinery?  Would making `self.param_scope = nn.ParamDict()` solve your issue of `pyro.param` being undefined in this new setting?

Reply:
I added a persistent parameter scope state to the `pyro.nn._Context` class, it seems to work as expected.